### PR TITLE
Batched fetching

### PR DESCRIPTION
**Pull request by KnVerey**
## Problem

Every polling loop, including the one to check the initial status, makes an API call for every single resource in the set. When you've got over a thousand resources, that takes a long time, even with concurrency!

## Solution

Make one request per resource type in the set instead, with a fallback if there's a cache miss for a type a resource instance needs.

We tested this in our canary environment and saw the "checking initial status" step decrease from 24s to 6s. The `apply` + initial polling loop step also decreased from 44s to 24s (probably because `apply --prune` is still slow). I would expect this time to grow somewhat with larger resource sets (raw `kubectl get` is slower on larger clusters), but no longer linearly with the number of resources.

A nice side-effect of this design is that we can remove the `kubectl` access from the instances entirely, preventing people from using it when they shouldn't be (i.e. outside polling loops).

**Reviewers**, please especially look for any concurrency gotchas we might have here. This new cache is effectively shared state in a multi-threaded environment. Should we freeze some of it, for example?

## Still to do

- [x] Add unit tests for the new class

### CODE REVIEW

File: lib/kubernetes-deploy/sync_mediator.rb [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    def initialize(namespace:, context:, logger:)
+      @namespace = namespace
+      @context = context
+      @logger = logger
+      clear_cache
+    end
+
+    def get_instance(kind, resource_name)
+      if @cache.key?(kind)
+        @cache[kind].find { |r| r.dig("metadata", "name") == resource_name } || {}
```
Comment (dturn) [inline lib/kubernetes-deploy/sync_mediator.rb]: What do you think about the cache being a 2d array by kind and then name? (e.g. `@cache[kind][resource_name] || {}`)

Comment (KnVerey) [reply lib/kubernetes-deploy/sync_mediator.rb]: I think it's a good idea; preprocessing it into that form up front should in theory be more efficient overall.

File: lib/kubernetes-deploy/kubernetes_resource/deployment.rb [hunk 2]
```diff
@@ -6,53 +6,37 @@ class Deployment < KubernetesResource
...
-        @progress_condition = nil
-        @progress_deadline = @definition['spec']['progressDeadlineSeconds']
-        @desired_replicas = -1
-        @max_unavailable = @definition.dig('spec', 'strategy', 'rollingUpdate', 'maxUnavailable')
-      end
+    SYNC_DEPENDENCIES = %w(Pod ReplicaSet)
+    def sync(mediator)
+      super
+      @latest_rs = exists? ? find_latest_rs(mediator) : nil
+      @server_version = mediator.kubectl.server_version
```
Comment (dturn) [inline lib/kubernetes-deploy/kubernetes_resource/deployment.rb]: Its ugly but we could store this as a class level var to prevent every deployment from having to make the api call. Might not be worth the optimization.

File: test/unit/sync_mediator_test.rb [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+
+    # Configmap is cached, so no new requests should happen here
+    mediator.kubectl.expects(:run).never
+    r = mediator.get_instance('FakeConfigMap', @fake_cm.name)
+    assert_equal @fake_cm.name, r.dig('metadata', 'name')
+    missing = mediator.get_instance('FakeConfigMap', 'does-not-exist')
+    assert_equal({}, missing)
+  end
+
+  def test_get_instance_does_not_populate_the_cache
```
Comment (dturn) [inline test/unit/sync_mediator_test.rb]: This test doesn't look finished, it also looks like its covered by the first test

File: lib/kubernetes-deploy/kubernetes_resource/bugsnag.rb [hunk 4]
```diff
@@ -1,33 +0,0 @@
-# frozen_string_literal: true
```
[the remark this answers is not in the document]
Comment (KnVerey) [reply lib/kubernetes-deploy/kubernetes_resource/bugsnag.rb]: This is a Shopify custom resource that we 🔥 in production months ago. It's effectively dead code, so I deleted rather than updated it.

File: test/unit/sync_mediator_test.rb [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+    assert_equal expected, mediator.get_all('FakeConfigMap')
+    assert_equal expected, mediator.get_all('FakeConfigMap') # cached
+    assert_equal expected, mediator.get_all('FakeConfigMap') # cached
+  end
+
+  def test_get_all_does_not_cache_error_result_from_kubectl
+    stub_kubectl_response('get', 'FakeConfigMap', *@params, success: false, resp: { "items" => [] }, err: 'no').times(2)
+    stub_kubectl_response('get', 'FakeConfigMap', @fake_cm.name, *@params, resp: @fake_cm.kubectl_response, times: 1)
+    assert_equal [], mediator.get_all('FakeConfigMap')
+    assert_equal [], mediator.get_all('FakeConfigMap', "fake" => "false", "type" => "fakeconfigmap")
```
[the remark this answers is not in the document]
Comment (KnVerey) [reply test/unit/sync_mediator_test.rb]: Yeah, admittedly it's unlikely we'd introduce a behavioural difference there, but I thought it couldn't hurt. I can remove it if you want.

File: lib/kubernetes-deploy/kubernetes_resource/memcached.rb [hunk 6]
```diff
@@ -4,14 +4,16 @@ class Memcached < KubernetesResource
...
+      @deployment = mediator.get_instance(Deployment.kind, "memcached-#{@name}")
+      @service = mediator.get_instance(Service.kind, "memcached-#{@name}")
+      @configmap = mediator.get_instance(ConfigMap.kind, CONFIGMAP_NAME)
+    end
+
+    def status
+      if deployment_ready? && service_ready? && configmap_ready?
         "Provisioned"
       else
         "Unknown"
```
[the remark this answers is not in the document]
Comment (KnVerey) [reply lib/kubernetes-deploy/kubernetes_resource/memcached.rb]: That link isn't working for me, but that seems like a mistake. Good catch!

**Comment by KnVerey** [comment]
> I was wondering whether there is a reason why the code sometimes uses dig(a,b,c) and in other places [a][b][c]. Any particular reason?

We were originally using an older version of ruby on this project, and I have a vague memory of updating cases like `a[b][c] if a.key?(b)` to use `dig` when we upgraded, but not bothering beyond that. So probably a combination of legacy and inconsistency on my part. Feel free to call them out.

> Are we okay with only having tests for deployments and pods?

Since we're so reliant on correct interactions with multiple versions of k8s, for the most part, this project has focussed on maintaining excellent integration coverage. We typically add unit tests when there are edge cases that are difficult/impossible to get integration coverage for. If you see any code that isn't covered, please call it out and I'll add coverage one way or another. One exception is the classes for Shopify's custom resources, which are virtually uncovered. The work to make them dynamic will give us generic coverage for them, but until then it is difficult since our test clusters don't run the backing controllers. Typically I deploy my test app locally as a 🎩 before merging any PRs that touch those classes. That definitely sucks. Maybe now's the time to change it... we could have the test itself inject dummy resources into the namespace to at least cover the basics. See also: [codecov](https://codecov.io/gh/Shopify/kubernetes-deploy/pull/251)

File: test/unit/sync_mediator_test.rb (hunk 5 above)
[the remark this answers is not in the document]
Comment (klautcomputing) [reply test/unit/sync_mediator_test.rb]: Nope totally fine, just took me a second to figure out why you added the test. Maybe just add a comment to it so that no one else has to wonder why.

**Comment by KnVerey** [comment]
Updated and rebased.

**Comment by sirupsen** [comment]
(not reviewing but pretty excited about this)

**Comment by KnVerey** [comment]
🎩 with a test app with shopify CRs successful
